text/0000-edition-based-method-disambiguation.md

```diff
@@ -0,0 +1,202 @@
+  --> src/lib.rs:23:10
+   |
+23 |     it.intersperse(sep)
+   |        ^^^^^^^^^^^ multiple `intersperse` found
+   |
+note: candidate #1 is defined in an impl of the trait `Iterator` for the type `MyIter`
+  --> src/lib.rs:7:1
+   |
+7  | impl Iterator for MyIter {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^
+   = note: this was introduced in the current edition and has been deprioritized to prevent breakage
+   = warning: in the next edition this warning will become an error
+note: candidate #2 is defined in an impl of the trait `Itertools` for the type `MyIter`
+  --> src/lib.rs:16:5
+   |
+16 |     fn intersperse(self, separator: Self::Item) -> Intersperse<Self> {
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   = note: to avoid the ambiguity, this candidate was selected to prevent breakage in edition < 20XX
+help: disambiguate the associated function for candidate #1
+   |
+23 |     Iterator::intersperse(it, sep)
+   |
+help: disambiguate the associated function for candidate #2
+   |
+23 |     Itertools::intersperse(it, sep)
+   |
+```
+
+These errors are an expected stage of the standard library development
+lifecycle, where methods are experimented within 3rd party crates and then
+moved into the standard library once they've been thorougly tested. A classic
+example of this is the `Itertools` crate for experimenting with extentions to
+the `Iterator` trait. However this problem isn't restricted to extension traits
+of existing standard library traits, and can indeed become a problem whenever
+any two methods have the same name.
+
+You can fix issues like this by manually editing the code to select the
+specific version of the method you wish to use or, in certain common cases, you
+can use cargo fix. cargo fix will make assumptions about how the methods relate
+depending on if you're using cargo fix for an edition upgrade or not.
+
+* **Within same edition** cargo fix will assume that the new method is a drop
+  in replacement of the existing downstream one and will disambiguate by
+  selecting the upstream method defined in `std`.
+* **As part of an edition upgrade** cargo fix will prioritize maintaining the
+  same behavior, and will disambiguate by selecting the existing method that
+  was being used previously.
+
+To run cargo fix within the same edition run:
+
+```
+cargo fix
+```
+
+In the example above this would replace the ambiguous code with
+`Iterator::intersperse(it, sep)`, selecting the new implementation.
+
+To run cargo fix as part of an edition upgrade run:
+
+```
+cargo fix --edition
+```
+
+In the example above this would replace the ambiguous code with
+`Itertools::intersperse(it, sep)`, maintaining the existing behavior.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature will be implemented by modifying the `rustc_stable` and
+`rustc_const_stable` attributes to support an additional optional `edition`
```

Comment:
Can you elaborate on the need for the update to `rustc_const_stable`?  Is there a scenario where people want the const version of the function from a crate but the non-const version from std?

Having method resolution be different between `const fn` and `fn` seems surprising to me, so I might have this depend only on the `rustc_stable` attribute -- or even make a new one just for the edition.

Come to think of it, should this resolution logic also apply to *unstable*-but-feature-enabled methods on nightly?

Reply:
I only added `rustc_const_stable` for completeness with the assumption that the method resolution behavior was the same regardless of constness, I didn't really think out the implications and didn't realize that this was referring to "stable-non-const vs unstable-const". I definitely didn't intend to suggest that the method resolution behavior should differ, I meant to do exactly the opposite, lol. I'll remove the reference to `rustc_const_stable`.

Reply:
Wait actually, I just checked the docs again and I'm confused. [`rustc_const_unstable`](https://rustc-dev-guide.rust-lang.org/stability.html#rustc_const_unstable) is the one that handles unstably const stuff but [`rustc_const_stable`](https://rustc-dev-guide.rust-lang.org/stability.html#rustc_const_stable) seems to be for const that is stable. I'm not sure exactly what the difference between `rustc_const_stable` and `rustc_stable`. I'm guessing it's just for tracking when constness was stabilized? I don't think the stabilization of constness would matter here, so long as the method exists at all I'm assuming it would cause name conflicts, so I'm think adding an `edition` field to does nothing.

Reply:
> I'm guessing it's just for tracking when constness was stabilized.

That's right.  I agree it shouldn't matter for name resolution.